src/main/java/org/opensearch/plugin/insights/core/service/categorizer/QueryShapeGenerator.java

```diff
@@ -74,9 +34,9 @@
 public class QueryShapeGenerator {
     static final String EMPTY_STRING = "";
     static final String ONE_SPACE_INDENT = " ";
-    static final Map<Class<?>, List<Function<Object, String>>> QUERY_FIELD_DATA_MAP = FieldDataMapHelper.getQueryFieldDataMap();
-    static final Map<Class<?>, List<Function<Object, String>>> AGG_FIELD_DATA_MAP = FieldDataMapHelper.getAggFieldDataMap();
-    static final Map<Class<?>, List<Function<Object, String>>> SORT_FIELD_DATA_MAP = FieldDataMapHelper.getSortFieldDataMap();
+    static final ConcurrentHashMap<Class<?>, List<Method>> QUERY_FIELD_DATA_MAP = FieldDataMapHelper.getQueryFieldDataMap();
+    static final ConcurrentHashMap<Class<?>, List<Method>> AGG_FIELD_DATA_MAP = FieldDataMapHelper.getAggFieldDataMap();
+    static final ConcurrentHashMap<Class<?>, List<Method>> SORT_FIELD_DATA_MAP = FieldDataMapHelper.getSortFieldDataMap();
```

Comment:
Do we really need `ConcurrentHashMap`? Also, can we not eliminate the map altogether and check for `isinstanceof WithFieldName` ?

Reply:
Using `ConcurrentHashMap` to eliminate need to check for null keys/values. 

We can use a simple `if` check today, but the maps allows for better extensibility. Assuming we still have other field data in the potential roadmap, I think we can keep the maps.

Reply:
IMO, having this static Map is not really adding any value. Why do you think the map approach is more extensible?

Reply:
Is `ConcurrentHashMap` only used to prevent null keys? IMHO if there's no concurrent access cases it might not be worth doing it.

Reply:
Agreed. Lets skip `ConcurrentHashMap` if this is the case

Reply:
@jainankitk Every time we add a new type of builder data, it can potentially be another `if builder instanceof X ...` check.

@ansjcy @deshsidd These maps are read only. I believe ConcurrentHashMap and HashMap are similar for reads as synchronization is only on writes.

Reply:
> @ansjcy @deshsidd These maps are read only. I believe ConcurrentHashMap and HashMap are similar for reads as synchronization is only on writes.

Yeah, I am not concerned about `ConcurrentHashMap` since we are writing to this `Map` only during the class initialization. Although, I am questioning if really need to keep it? The `instanceof` check is similar to the `map.contains` check?